### DISCUSSION TITLE
Moving from VMSS Agent to Containers

### DISCUSSION THREAD

**Discussion by thepaulmacca** [actions/virtual-environments#3002] [Ideas]
Sorry there wasn't a better category to put this under...

I'm looking to move away from Self-Hosted VMSS Agents to Containerized Agents. Can anyone shed some light on how the MS Hosted Docker images are built? I understand they were archived quite some time ago, but seeing as MS use container agents, how are these images getting built? I can't imagine there's a single Dockerfile, as there's a lot involved in building the VM images

Any pointers appreciated. Thanks!

**Comment by catthehacker** [comment]
VM images are built using Packer, not Docker. It's full virtual machine disk that is built. The instructions are here: https://github.com/actions/virtual-environments/blob/main/help/CreateImageAndAzureResources.md

From [Azure Pipelines documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/agents):
> If your pipelines are in Azure Pipelines, then you've got a convenient option to run your jobs using a Microsoft-hosted agent. With Microsoft-hosted agents, maintenance and upgrades are taken care of for you. Each time you run a pipeline, you get a fresh virtual machine. The virtual machine is discarded after one use. Microsoft-hosted agents can run jobs [directly on the VM](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops) or [in a container](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/container-phases?view=azure-devops).

**Reply by thepaulmacca** [reply]
Fully aware that VM Images are built using Packer, but it was my understanding that the MS hosted agents are actually Containers and not VMs. Have I got the wrong end of the stick?

**Reply by miketimofeev** [reply]
No, we don't use containers for Hosted agents at the moment

**Reply by thepaulmacca** [reply]
Ah right, I was under the impression that you were. No worries, thanks for confirming

**Reply by catthehacker** [reply]
Pretty sure it's just VM without any containerisation in between ADO/GHA and user since those are not shared thus there is no need to run stuff in Docker (which itself could cause massive issues with runtimes).

**Reply by thepaulmacca** [reply]
Is there any way that we can mark these discussions as resolved like they do in the bicep repo? I don't see the option here, but would help a lot I think

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
